api/buckets/service.go

```diff
@@ -290,126 +289,194 @@ func (s *Service) newDirFromExistingPath(ctx context.Context, pth path.Path, key
-			if err != nil {
-				return nil, err
-			}
-			leaves = append(leaves, ltn)
-		}
-		if !isLeaf {
-			// We're done with these links, wrap up the new dir
-			dir := unixfs.EmptyDirNode()
-			dir.SetCidBuilder(dag.V1CidPrefix())
-			for _, l := range leaves {
-				if err := dir.AddNodeLink(l.name, l.node); err != nil {
-					return nil, err
+// This method returns a map of all nodes keyed by their _original_ plaintext cid.
+func (s *Service) encryptDag(ctx context.Context, ds ipld.DAGService, root ipld.Node, key []byte, add *namedNode) (map[cid.Cid]*namedNode, error) {
+	// Step 1: Create a preordered list of joint and leaf nodes
+	var stack, joints, leaves []*namedNode
+	stack = append(stack, &namedNode{node: root})
+	var cur *namedNode
+
+	for len(stack) > 0 {
+		n := len(stack) - 1
+		cur = stack[n]
+		stack = stack[:n]
+	types:
+		switch cur.node.(type) {
+		case *dag.RawNode:
+			leaves = append(leaves, cur)
+		case *dag.ProtoNode:
+			// Add links to the stack
+			for _, l := range cur.node.Links() {
+				if l.Name == "" {
+					leaves = append(leaves, cur)
+					break types // We have discovered a raw node wrapper
 				}
-			}
-			if add != nil {
-				if err := dir.AddNodeLink(add.name, add.node); err != nil {
+				ln, err := l.GetNode(ctx, ds)
+				if err != nil {
 					return nil, err
 				}
-				tree[add.node.Cid()] = add
+				stack = append(stack, &namedNode{name: l.Name, node: ln})
 			}
-			cn, err := encryptNode(dir, key)
-			if err != nil {
-				return nil, err
+			joints = append(joints, cur)
+		default:
+			return nil, errInvalidNodeType
+		}
+	}
+
+	// Step 2: Encrypt all leaf nodes in parallel
+	nmap := newNamedNodes()
+	var wg sync.WaitGroup
+	errs := make(chan error)
+	wgDone := make(chan bool)
+	for _, l := range leaves {
+		wg.Add(1)
+		go func(l *namedNode) {
+			defer wg.Done()
+			var cn ipld.Node
+			switch l.node.(type) {
+			case *dag.RawNode:
+				data, err := encryptData(l.node.RawData(), key)
+				if err != nil {
+					errs <- err
+					return
+				}
+				cn = dag.NewRawNode(data)
+			case *dag.ProtoNode:
+				var err error
+				cn, err = s.encryptFileNode(ctx, l.node, key)
+				if err != nil {
+					errs <- err
+					return
+				}
 			}
-			tn = &treeNode{
-				name: name,
+			nmap.Store(l.node.Cid(), &namedNode{
+				name: l.name,
 				node: cn,
-			}
-			break
-		}
+			})
+		}(l)
+	}
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-errs:
+		close(errs)
+		return nil, err
+	}
 
-		// Encrypt and add the file node
-		fn, err := s.IPFSClient.Unixfs().Get(ctx, path.IpfsPath(root.Cid()))
-		if err != nil {
-			return nil, err
-		}
-		defer fn.Close()
-		file := ipfsfiles.ToFile(fn)
-		if file == nil {
-			return nil, fmt.Errorf("node is a directory")
-		}
-		r, err := dcrypto.NewEncrypter(file, key)
-		if err != nil {
-			return nil, err
+	// Step 3: Encrypt joint nodes in reverse, walking up to root
+	for i := len(joints) - 1; i >= 0; i-- {
```

Comment:
So this works since the iterative process is still depth-first, so encrypted nodes should have executed in the right order to always find the encrypted links, and so on.

Reply:
Yep, exactly.